File: 904. Fruit Into Baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int ans = 0;
        int count = 0;
        int[] freq = new int[fruits.length];
        int left = 0;
        for(int right = 0; right < fruits.length; right++){
            freq[fruits[right]]++;
            if(freq[fruits[right]] == 1) count++;

            while(count > 2){
                freq[fruits[left]]--;
                if(freq[fruits[left]] == 0) count--;
                left++;
            } 

            ans = Math.max(ans, right - left + 1);
        }

        return ans;
    }
}

File: 904. Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> hm;
        int left = 0, maxLen = 0;

        for (int right = 0; right < fruits.size(); ++right) {
            hm[fruits[right]]++;

            while (hm.size() > 2) {
                hm[fruits[left]]--;
                if (hm[fruits[left]] == 0) {
                    hm.erase(fruits[left]);
                }
                ++left;
            }

            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};